File: lib/services/exam_set_service.dart
import '../core/enums/app_enums.dart';
import '../core/utils/exam_period_time_helper.dart';
import '../models/exam/exam_period_assignment.dart';
import '../models/exam/exam_session.dart';
import '../models/exam/exam_set.dart';
import '../repositories/exam_period_assignment_repository.dart';
import '../repositories/exam_session_repository.dart';
import '../repositories/exam_set_repository.dart';

/// Item trong danh sách kỳ thi — tương đương `AssignedExamItemViewModel`.
class AssignedExamItem {
  final ExamPeriodAssignment assignment;
  final ExamSet? examSet;

  const AssignedExamItem({required this.assignment, this.examSet});
}

/// Chi tiết kỳ thi — tương đương `ExamSetDetailsViewModel`.
class ExamSetDetails {
  final ExamPeriodAssignment assignment;
  final ExamSet? examSet;
  final List<ExamSession> inProgressSessions;
  final bool canStartExam;
  final bool canContinueExam;
  final String periodStatusName;
  final String displayStatusName;

  const ExamSetDetails({
    required this.assignment,
    this.examSet,
    this.inProgressSessions = const [],
    this.canStartExam = false,
    this.canContinueExam = false,
    this.periodStatusName = '',
    this.displayStatusName = '',
  });
}

/// Dịch vụ kỳ thi — port từ `ExamSetService.cs`.
class ExamSetService {
  ExamSetService(
    this._examSetRepo,
    this._sessionRepo,
    this._assignmentRepo,
  );

  final ExamSetRepository _examSetRepo;
  final ExamSessionRepository _sessionRepo;
  final ExamPeriodAssignmentRepository _assignmentRepo;

  Future<({List<AssignedExamItem> items, String? error})> getIndex() async {
    final result = await _assignmentRepo.getMyAssignments();
    if (!result.success) {
      return (items: <AssignedExamItem>[], error: result.message);
    }

    final items = <AssignedExamItem>[];
    for (final assignment in result.data ?? []) {
      ExamSet? examSet;
      if (assignment.examSetId != null && assignment.examSetId! > 0) {
        final examSetResult =
            await _examSetRepo.getById(assignment.examSetId!);
        if (examSetResult.success &&
            examSetResult.data != null &&
            examSetResult.data!.status == 1) {
          examSet = examSetResult.data;
        }
      }
      items.add(AssignedExamItem(assignment: assignment, examSet: examSet));
    }

    _sortIndexItems(items);

    return (items: items, error: null);
  }

  static bool isAssignmentCompleted(ExamPeriodAssignment assignment) =>
      assignment.status == ExamPeriodAssignmentStatus.completed;

  static void _sortIndexItems(List<AssignedExamItem> items) {
    items.sort((a, b) {
      final aDone = isAssignmentCompleted(a.assignment);
      final bDone = isAssignmentCompleted(b.assignment);
      if (aDone != bDone) return aDone ? 1 : -1;

      if (!aDone) {
        final aStart = a.assignment.examPeriodStartAt;
        final bStart = b.assignment.examPeriodStartAt;
        if (aStart == null && bStart == null) return 0;
        if (aStart == null) return 1;
        if (bStart == null) return -1;
        return aStart.compareTo(bStart);
      }

      final aTime = a.assignment.sessionFinishedAt ??
          a.assignment.examPeriodEndAt ??
          a.assignment.examPeriodStartAt;
      final bTime = b.assignment.sessionFinishedAt ??
          b.assignment.examPeriodEndAt ??
          b.assignment.examPeriodStartAt;
      if (aTime == null && bTime == null) return 0;
      if (aTime == null) return 1;
      if (bTime == null) return -1;
      return bTime.compareTo(aTime);
    });
  }

  Future<({ExamSetDetails? model, String? error})> getDetails({
    required int assignmentId,
    required int userId,
  }) async {
    final assignmentsResult = await _assignmentRepo.getMyAssignments();
    if (!assignmentsResult.success) {
      return (
        model: null,
        error: assignmentsResult.message.isNotEmpty
            ? assignmentsResult.message
            : 'Không thể tải phân công.',
      );
    }

    final assignment = (assignmentsResult.data ?? [])
        .where((x) => x.id == assignmentId)
        .firstOrNull;

    if (assignment == null) {
      return (model: null, error: 'Bạn chưa được phân công kỳ thi này.');
    }

    ExamSet? examSet;
    if (assignment.examSetId != null && assignment.examSetId! > 0) {
      final examSetResult = await _examSetRepo.getById(assignment.examSetId!);
      if (examSetResult.success) examSet = examSetResult.data;
    }

    var inProgress = <ExamSession>[];
    if (assignment.status != ExamPeriodAssignmentStatus.completed &&
        assignment.status != ExamPeriodAssignmentStatus.cancelled &&
        ExamPeriodTimeHelper.canContinueExam(assignment)) {
      final inProgressResult = await _sessionRepo.getList(
        pageIndex: 1,
        pageSize: 10,
        userId: userId,
        status: ExamSessionStatus.inProgress,
      );
      inProgress = (inProgressResult.data ?? [])
          .where(
            (s) =>
                s.examType == assignment.examType &&
                (assignment.examSetId == null ||
                    assignment.examSetId! <= 0 ||
                    s.examSetId == assignment.examSetId),
          )
          .toList();
    }

    return (
      model: ExamSetDetails(
        assignment: assignment,
        examSet: examSet,
        inProgressSessions: inProgress,
        canStartExam: ExamPeriodTimeHelper.canStartExam(assignment),
        canContinueExam: ExamPeriodTimeHelper.canContinueExam(assignment),
        periodStatusName:
            ExamPeriodTimeHelper.examPeriodStatusName(assignment),
        displayStatusName: ExamPeriodTimeHelper.displayStatusName(assignment),
      ),
      error: null,
    );
  }
}

extension _FirstOrNull<E> on Iterable<E> {
  E? get firstOrNull {
    final iterator = this.iterator;
    if (iterator.moveNext()) return iterator.current;
    return null;
  }
}

File: ADMIN/Example-API-Flutter/Example.UserService.WebApp/Services/ExamSetService.cs
using Example.Common.Enums;
using Example.Common.Utilities;
using Example.UserService.API.Models;
using Example.UserService.WebApp.Helpers;
using Example.UserService.WebApp.Models.ExamSet;
using Example.UserService.WebApp.Repository.IRepository;
using Example.UserService.WebApp.Services.IServices;

namespace Example.UserService.WebApp.Services
{
    /// <summary>
    /// Dịch vụ danh sách kỳ thi được phân công: gộp dữ liệu phân công + bộ đề + session đang làm,
    /// và quyết định hiển thị nút bắt đầu/tiếp tục theo khung giờ kỳ thi và trạng thái phân công.
    /// </summary>
    public class ExamSetService : IExamSetService
    {
        private readonly IExamSetRepository _examSetRepository;
        private readonly IExamSessionRepository _examSessionRepository;
        private readonly IExamPeriodAssignmentRepository _assignmentRepository;

        // Tiêm ba repository: phân công, bộ đề và phiên thi để gộp dữ liệu trang index/chi tiết.
        public ExamSetService(
            IExamSetRepository examSetRepository,
            IExamSessionRepository examSessionRepository,
            IExamPeriodAssignmentRepository assignmentRepository)
        {
            _examSetRepository = examSetRepository;
            _examSessionRepository = examSessionRepository;
            _assignmentRepository = assignmentRepository;
        }

        /// <summary>Trang index: liệt kê mọi phân công của thí sinh kèm thông tin bộ đề (nếu đã gán).</summary>
        public async Task<ExamSetIndexViewModel> GetIndexAsync()
        {
            var assignmentsResult = await _assignmentRepository.GetMyAssignmentsAsync();
            if (!assignmentsResult.Success)
            {
                return new ExamSetIndexViewModel { ErrorMessage = assignmentsResult.Message };
            }

            var assignments = assignmentsResult.Data ?? new List<ExamPeriodAssignmentModel>();
            var items = new List<AssignedExamItemViewModel>();

            foreach (var assignment in assignments)
            {
                // Bộ đề có thể null (random khi start) — chỉ load nếu đã gán sẵn ExamSetId.
                ExamSetModel? examSet = null;
                if (assignment.ExamSetId > 0)
                {
                    var examSetResult = await _examSetRepository.GetByIdAsync(assignment.ExamSetId.Value);
                    // Quy tắc nghiệp vụ: chỉ hiển thị bộ đề đang Active trên danh sách.
                    if (examSetResult.Success && examSetResult.Data != null
                        && examSetResult.Data.Status == (int)StatusEnum.Active)
                    {
                        examSet = examSetResult.Data;
                    }
                }

                items.Add(new AssignedExamItemViewModel
                {
                    Assignment = assignment,
                    ExamSet = examSet
                });
            }

            return new ExamSetIndexViewModel { Items = items };
        }

        /// <summary>
        /// Trang chi tiết kỳ thi: xác minh phân công thuộc user, load bộ đề và session InProgress phù hợp.
        /// </summary>
        public async Task<(ExamSetDetailsViewModel? Model, string? ErrorMessage)> GetDetailsAsync(long assignmentId, long userId)
        {
            var assignmentsResult = await _assignmentRepository.GetMyAssignmentsAsync();
            if (!assignmentsResult.Success)
                return (null, assignmentsResult.Message ?? "Không thể tải phân công.");

            // Quy tắc nghiệp vụ: chỉ hiển thị phân công nằm trong danh sách của user hiện tại (chống truy cập chéo).
            var assignment = (assignmentsResult.Data ?? new List<ExamPeriodAssignmentModel>())
                .FirstOrDefault(x => x.Id == assignmentId);

            if (assignment == null)
                return (null, "Bạn chưa được phân công kỳ thi này.");

            ExamSetModel? examSet = null;
            if (assignment.ExamSetId > 0)
            {
                var examSetResult = await _examSetRepository.GetByIdAsync(assignment.ExamSetId.Value);
                if (examSetResult.Success && examSetResult.Data != null)
                    examSet = examSetResult.Data;
            }

            // Tìm session InProgress cùng loại thi (và cùng bộ đề nếu phân công đã gán sẵn).
            var inProgressResult = await _examSessionRepository.GetListAsync(new ExamSessionSearchModel
            {
                PageIndex = 1,
                PageSize = 10,
                UserId = userId,
                Status = (int)ExamSessionStatusEnum.InProgress
            });
            var inProgress = new List<ExamSessionModel>();
            // Quy tắc nghiệp vụ: chỉ cho tiếp tục khi phân công chưa hoàn thành và còn trong khung giờ thi.
            if (assignment.Status != (int)ExamPeriodAssignmentStatusEnum.Completed
                && assignment.Status != (int)ExamPeriodAssignmentStatusEnum.Cancelled
                && ExamPeriodTimeHelper.CanContinueExam(assignment))
            {
                inProgress = (inProgressResult.Data ?? new List<ExamSessionModel>())
                    .Where(s => s.ExamType == assignment.ExamType
                        && (assignment.ExamSetId == null || assignment.ExamSetId <= 0 || s.ExamSetId == assignment.ExamSetId))
                    .ToList();
            }

            return (new ExamSetDetailsViewModel
            {
                Assignment = assignment,
                ExamSet = examSet,
                InProgressSessions = inProgress,
                CanStartExam = CanStartExam(assignment),
                CanContinueExam = CanContinueExam(assignment),
                PeriodStatusName = GetExamPeriodStatusName(assignment),
                DisplayStatusName = GetDisplayStatusName(assignment)
            }, null);
        }

        /// <summary>Ánh xạ trạng thái khung giờ kỳ thi sang nhãn hiển thị cho thí sinh.</summary>
        public string GetExamPeriodStatusName(ExamPeriodAssignmentModel assignment)
        {
            var availability = ExamPeriodTimeHelper.GetAvailability(assignment);
            return availability switch
            {
                ExamPeriodAvailability.NotYetStarted => "Đã công bố",
                ExamPeriodAvailability.Open => "Đã công bố",
                ExamPeriodAvailability.Ended => "Đã đóng",
                _ => DisplayHelper.EnumName<ExamPeriodStatusEnum>(assignment.ExamPeriodStatus)
            };
        }

        /// <summary>Nhãn trạng thái tổng hợp: ưu tiên hết giờ/chưa đến giờ trước trạng thái phân công.</summary>
        public string GetDisplayStatusName(ExamPeriodAssignmentModel assignment)
        {
            var availability = ExamPeriodTimeHelper.GetAvailability(assignment);
            if (availability == ExamPeriodAvailability.Ended)
                return "Kết thúc";
            if (availability == ExamPeriodAvailability.NotYetStarted)
                return "Chưa đến giờ thi";
            return GetAssignmentStatusName(assignment.Status);
        }

        /// <summary>Quy tắc: được bắt đầu thi khi khung giờ mở và phân công cho phép.</summary>
        public bool CanStartExam(ExamPeriodAssignmentModel assignment)
            => ExamPeriodTimeHelper.CanStartExam(assignment);

        /// <summary>Quy tắc: được tiếp tục khi còn trong khung giờ và chưa hoàn thành phân công.</summary>
        public bool CanContinueExam(ExamPeriodAssignmentModel assignment)
            => ExamPeriodTimeHelper.CanContinueExam(assignment);

        /// <summary>Chuyển mã loại thi (Trial/Real) sang tên tiếng Việt.</summary>
        public string GetExamTypeName(int examType) => examType switch
        {
            (int)ExamTypeEnum.Trial => "Thi thử",
            (int)ExamTypeEnum.Real => "Thi thật",
            _ => DisplayHelper.EnumName<ExamTypeEnum>(examType)
        };

        /// <summary>Chuyển mã trạng thái phân công sang nhãn hiển thị.</summary>
        public string GetAssignmentStatusName(int status) => status switch
        {
            (int)ExamPeriodAssignmentStatusEnum.Assigned => "Đã phân công",
            (int)ExamPeriodAssignmentStatusEnum.InProgress => "Đang thi",
            (int)ExamPeriodAssignmentStatusEnum.Completed => "Hoàn thành",
            (int)ExamPeriodAssignmentStatusEnum.Absent => "Vắng thi",
            (int)ExamPeriodAssignmentStatusEnum.Cancelled => "Bị hủy bỏ",
            _ => DisplayHelper.EnumName<ExamPeriodAssignmentStatusEnum>(status)
        };
    }
}
